process/interceptors/peerAuthenticationInterceptor.go

```diff
@@ -0,0 +1,116 @@
+	"github.com/ElrondNetwork/elrond-go/process"
+)
+
+// ArgPeerAuthenticationInterceptor is the argument for the peer authentication interceptor
+type ArgPeerAuthenticationInterceptor struct {
+	ArgSingleDataInterceptor
+	Marshalizer             marshal.Marshalizer
+	ValidatorChecker        process.ValidatorChecker
+	AuthenticationProcessor process.PeerAuthenticationProcessor
+}
+
+type peerAuthenticationInterceptor struct {
+	*baseDataInterceptor
+	validatorChecker            process.ValidatorChecker
+	peerAuthenticationProcessor process.PeerAuthenticationProcessor
+}
+
+// NewPeerAuthenticationInterceptor hooks a new interceptor for packed multi data containing peer authentication instances
+func NewPeerAuthenticationInterceptor(arg ArgPeerAuthenticationInterceptor) (*peerAuthenticationInterceptor, error) {
+	err := checkArguments(arg.ArgSingleDataInterceptor)
+	if err != nil {
+		return nil, err
+	}
+	if check.IfNil(arg.Marshalizer) {
+		return nil, process.ErrNilMarshalizer
+	}
+	if check.IfNil(arg.ValidatorChecker) {
+		return nil, process.ErrNilValidatorChecker
+	}
+	if check.IfNil(arg.AuthenticationProcessor) {
+		return nil, process.ErrNilAuthenticationProcessor
+	}
+
+	interceptor := &peerAuthenticationInterceptor{
+		baseDataInterceptor: &baseDataInterceptor{
+			throttler:        arg.Throttler,
+			antifloodHandler: arg.AntifloodHandler,
+			topic:            arg.Topic,
+			currentPeerId:    arg.CurrentPeerId,
+			processor:        arg.Processor,
+			debugHandler:     resolver.NewDisabledInterceptorResolver(),
+			marshalizer:      arg.Marshalizer,
+			factory:          arg.DataFactory,
+		},
+		validatorChecker:            arg.ValidatorChecker,
+		peerAuthenticationProcessor: arg.AuthenticationProcessor,
+	}
+
+	return interceptor, nil
+}
+
+// ProcessReceivedMessage is the callback func from the p2p.Messenger and will be called each time a new message was received
+// (for the topic this validator was registered to)
+func (pai *peerAuthenticationInterceptor) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedPeer core.PeerID) error {
+	multiDataBuff, err := pai.preProcessMessage(message, fromConnectedPeer)
+	if err != nil {
+		return err
+	}
+
+	observerKeyFound := false
+	for _, dataBuff := range multiDataBuff {
+		var interceptedData process.InterceptedData
+		interceptedData, err = pai.interceptedData(dataBuff, message.Peer(), fromConnectedPeer)
+		if err != nil {
+			pai.throttler.EndProcessing()
+			return err
+		}
+
+		peerAuth, ok := interceptedData.(process.InterceptedPeerAuthentication)
+		if !ok {
+			//intercepted data is not of type interceptedPeerInfo
+			cause := "intercepted data is not of type process.InterceptedPeerInfo"
+			pai.blackListPeers(cause, nil, message.Peer(), fromConnectedPeer)
+			pai.throttler.EndProcessing()
+
+			return errors.New(cause)
+		}
+
+		var shardID uint32
+		_, shardID, err = pai.validatorChecker.GetValidatorWithPublicKey(peerAuth.PublicKey())
```

Comment:
are we sure that an error can be returned only if it is an observer? Maybe make the condition more specific, like `err == ErrPubKeyIsNotAValidator`

Reply:
I would keep this as it is for one single reason: p2p package will depend on the sharding package for that type of error check. Could have also redefined the error but that would have caused some unexpected behavior (hard to catch in integration tests) when the indexedHashedCoordinator would have changed that particular error.

Reply:
ok